woocommerce/woocommerce-ios
# [Manage Categories] Build and Render Category Selector tree view

References https://github.com/woocommerce/woocommerce-ios/issues/2000

# Why

This PR takes the full category list implemented in #2130 and sort it in a way that allow us to show sub-categories within parent categories  in a simple table view screen.

![category tree](https://user-images.githubusercontent.com/562080/79278379-32fbfa00-7e71-11ea-8bc3-6124677b2961.png)


# How

## Tree Generation

This PR introduces a type called `ProductCategoryViewModelBuilder` which is in charge of taking a unsorted list of categories and output a sorted list of `ProductCategoryViewModel`s.

The underlying algorithm takes the list of categories and builds a `Tree` structure that uses `Dictionary<CategoryID, ArrayOfSubcategories>` as an underlying data structure. This `Tree` will help us finding each immediate sub-category of a given category.

Then the `ProductCategoryViewModelBuilder` traverses the `Tree` doing a standard `Depth First Search` in order to get all categories that correspond to a given root category. A root category is a category whose `parentID` is `0`.

I've decided to do the tree `DFS` and the `ViewModel` creation in the same pass to avoid an extra iteration as I don't know how many categories a site can contain and wanted to be on the safe side 😇 

## ViewModel
Now, after synchronizing categories, the `ProductCategoryListViewModel` takes all fetched objects and converts them into `ProductCategoryViewModel`s  and exposes them through the `categoryViewModels` property to be consumed bu the `ViewController`

## Cell UI
Also, I've modified the `ProductCategoryTableViewCell` to properly render the `ViewModel` indentation level.

# Testing Steps
- Log into a site that has multiple categories with sub categories
- Go to the products tab
- Tap on a product
- Tap on the categories field
- See that categories are rendered in a tree structure
